### PR TITLE
PHP 3.0

### PR DESCRIPTION
**Pull request by jrossiter**
As per ongoing discussion with @sougou. 

Initial commit.

### CODE REVIEW

File: php/src/Vitess/VTGateConn.php [hunk 1]
```diff
@@ -17,6 +19,11 @@ class VTGateConn
     protected $keyspace;
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete this?

**Comment by jrossiter** [comment]
I signed it!

File: php/src/Vitess/VTGateConn.php [hunk 2]
```diff
@@ -36,27 +43,32 @@ class VTGateConn
...
         if ($ctx->getCallerId()) {
             $request->setCallerId($ctx->getCallerId());
         }
         $response = $this->client->execute($ctx, $request);
+        $this->session = $response->getSession();
         ProtoUtils::checkError($response);
         return new Cursor($response->getResult());
     }
 
     public function executeShards(Context $ctx, $query, $keyspace, array $shards, array $bind_vars, $tablet_type)
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete.

File: php/src/Vitess/VTGateConn.php [hunk 3]
```diff
@@ -72,6 +84,7 @@ public function executeShards(Context $ctx, $query, $keyspace, array $shards, ar
     public function executeKeyspaceIds(Context $ctx, $query, $keyspace, array $keyspace_ids, array $bind_vars, $tablet_type)
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete.

File: php/src/Vitess/VTGateConn.php [hunk 4]
```diff
@@ -87,6 +100,7 @@ public function executeKeyspaceIds(Context $ctx, $query, $keyspace, array $keysp
     public function executeKeyRanges(Context $ctx, $query, $keyspace, array $key_ranges, array $bind_vars, $tablet_type)
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete.

File: php/src/Vitess/VTGateConn.php [hunk 5]
```diff
@@ -102,6 +116,7 @@ public function executeKeyRanges(Context $ctx, $query, $keyspace, array $key_ran
     public function executeEntityIds(Context $ctx, $query, $keyspace, $entity_column_name, array $entity_keyspace_ids, array $bind_vars, $tablet_type)
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete.

File: php/src/Vitess/VTGateConn.php [hunk 6]
```diff
@@ -125,6 +140,7 @@ public function executeEntityIds(Context $ctx, $query, $keyspace, $entity_column
     public function executeBatchShards(Context $ctx, array $bound_shard_queries, $tablet_type, $as_transaction)
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete.

File: php/src/Vitess/VTGateConn.php [hunk 7]
```diff
@@ -150,6 +166,7 @@ public function executeBatchShards(Context $ctx, array $bound_shard_queries, $ta
     public function executeBatchKeyspaceIds(Context $ctx, array $bound_keyspace_id_queries, $tablet_type, $as_transaction)
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: delete.

File: php/src/Vitess/VTGateConn.php [hunk 8]
```diff
@@ -11,11 +13,10 @@ class VTGateConn
...
 
     /**
-     *
-     * @var string The connection-wide keyspace for execute() calls.
+     * @var Session
      */
-    protected $keyspace;
-
+    protected $session;
+
```
Comment (sougou) [inline php/src/Vitess/VTGateConn.php]: extra blanks.

**Comment by sougou** [comment]
Actually, we should also have @Rastusik review this. We have to make sure we don't break anything for him :).

**Comment by jrossiter** [comment]
Sure no problem. In the meantime I'll try it internally as well.

**Comment by Rastusik** [comment]
what are we waiting for on this one? I don't know if I should adapt the PDO driver

**Comment by sougou** [comment]
We were waiting for you to get back from vacation.
@jrossiter If this works for you, I can submit it (we don't know how to test this ourselves yet).

**Comment by jrossiter** [comment]
@Rastusik The furthest I could get with testing the changes was:

```Notice: SQL Query failed: 0 2: vtgate: http://vtgate-test-krflf:15001/: paramsUnsharded: No partition found for tabletType unknown in keyspace test_keyspace in .../Vitess.php on line 548```

I provided a blank keyspace in the VTGateConn constructor and any queries would come up with the above error. I was waiting for you to get back from vacation for some advice on how to resolve the above issue. 

Is the TabletType supposed to be inferred by the keyspace?

**Comment by sougou** [comment]
That sounds like an error you would get if you sent the request to an older vtgate. If you've verified that it's the newer vtgate, we may have to dig in to trouble-shoot. It means that the target_string is somehow not reaching the vtgate.

**Comment by jrossiter** [comment]
I'll get back to you shortly.

**Comment by jrossiter** [comment]
We've been able to test that the branch works against the latest vtgate.

**Comment by sougou** [comment]
LGTM

**Pull request merged**